src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -53,6 +54,42 @@ impl CommandRunner for SpeculatingCommandRunner {
   }
 }
 
+impl CommandRunner for SpeculatingCommandRunner {
+  fn is_compatible_request(&self, req: &MultiPlatformExecuteProcessRequest) -> bool {
+    self.primary.is_compatible_request(req) || self.secondary.is_compatible_request(req)
+  }
+
+  fn get_compatible_request(
+    &self,
+    req: &MultiPlatformExecuteProcessRequest,
+  ) -> ExecuteProcessRequest {
+    match (
+      self.primary.is_compatible_request(req),
+      self.secondary.is_compatible_request(req),
+    ) {
+      (true, _) => self.primary.get_compatible_request(req),
+      (_, true) => self.secondary.get_compatible_request(req),
+      _ => panic!("No compatible requests found"),
+    }
+  }
+
+  fn run(
+    &self,
+    req: MultiPlatformExecuteProcessRequest,
+    workunit_store: WorkUnitStore,
+  ) -> BoxFuture<FallibleExecuteProcessResult, String> {
+    match (
+      self.primary.is_compatible_request(&req),
+      self.secondary.is_compatible_request(&req),
+    ) {
+      (true, true) => self.speculate(req, workunit_store),
+      (true, false) => self.primary.run(req, workunit_store),
+      (false, true) => self.secondary.run(req, workunit_store),
+      (false, false) => panic!("No platforms found to run the request"),
```

Comment:
Rather than panicing, this should `format!(...)` to return an error string, ideally with a bit of debug information about which platforms would be likely to work.

In general, removing almost all of the panics that are added by this review would be a good idea: either by erroring, or by returning an `Option` that the caller should handle eagerly (having `get_compatible_request` return an `Option` would handle the one above this).

Reply:
Makes total sense. Thanks.